File: backend/app/models/device.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Boolean, Text, Enum as SQLEnum, ForeignKey, Integer
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base
from app.core.types import UUID


class DeviceStatus(str, enum.Enum):
    PENDING = "pending"          # Waiting for enrollment
    ACTIVE = "active"            # Normal operation
    LOCKED = "locked"            # Device is locked
    WARNING = "warning"          # Warning issued
    WIPED = "wiped"              # Factory reset done
    DEACTIVATED = "deactivated"  # Removed from system


class Device(Base):
    __tablename__ = "devices"
    
    id = Column(UUID(), primary_key=True, default=uuid.uuid4)
    # IMEI can be null on Android 10+ without special permissions
    imei = Column(String(20), unique=True, nullable=True, index=True)
    imei2 = Column(String(20), index=True)
    serial_number = Column(String(100), index=True)
    
    # Persistent ID - survives factory reset (hash of hardware IDs)
    # This is the primary device identifier when IMEI is not available
    persistent_device_id = Column(String(64), unique=True, index=True)
    
    # Android ID - changes after factory reset
    android_id = Column(String(64))
    
    device_model = Column(String(100))
    manufacturer = Column(String(100))
    brand = Column(String(100))
    device_name = Column(String(100))
    product = Column(String(100))
    board = Column(String(100))
    hardware = Column(String(100))
    build_fingerprint = Column(String(255))
    
    android_version = Column(String(20))
    sdk_version = Column(String(10))
    
    # FCM Token for push notifications
    fcm_token = Column(Text)
    
    # Status
    status = Column(SQLEnum(DeviceStatus), default=DeviceStatus.PENDING, index=True)
    is_online = Column(Boolean, default=False)
    last_seen = Column(DateTime)
    
    # Track factory resets
    factory_reset_count = Column(Integer, default=0)
    last_factory_reset = Column(DateTime)
    
    # Security
    is_rooted = Column(Boolean, default=False)
    safety_net_passed = Column(Boolean)
    is_device_owner = Column(Boolean, default=False)
    is_admin_active = Column(Boolean, default=False)
    app_version = Column(String(20))
    
    # Enrollment
    enrollment_code = Column(String(50), unique=True)
    enrolled_at = Column(DateTime)
    enrolled_by = Column(UUID(), ForeignKey("users.id"))
    
    # GPS Tracking
    last_latitude = Column(String(50))
    last_longitude = Column(String(50))
    last_location_time = Column(DateTime)
    last_location_address = Column(Text)
    
    # Camera
    camera_active = Column(Boolean, default=False)
    last_photo_url = Column(Text)
    last_photo_time = Column(DateTime)
    
    # App visibility state
    is_app_hidden = Column(Boolean, default=False)
    is_app_disabled = Column(Boolean, default=False)
    
    # Real-time device info (updated every 2 seconds via heartbeat)
    battery_level = Column(Integer)
    is_charging = Column(Boolean)
    network_type = Column(String(20))
    
    # Relations
    customer_id = Column(UUID(), ForeignKey("customers.id"), index=True)
    customer = relationship("Customer", back_populates="devices")
    commands = relationship("DeviceCommand", back_populates="device")
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    def __repr__(self):
        return f"<Device {self.imei}>"

File: backend-node/src/models/device.js
import { randomUUID } from 'crypto';
import { DataTypes } from 'sequelize';
import { sequelize } from '../db.js';
import { enumField, DeviceStatus } from '../utils/enums.js';

export const Device = sequelize.define(
  'Device',
  {
    id: { type: DataTypes.STRING(36), primaryKey: true, defaultValue: randomUUID },
    imei: { type: DataTypes.STRING(20), allowNull: true, unique: true },
    imei2: { type: DataTypes.STRING(20) },
    serial_number: { type: DataTypes.STRING(100) },
    persistent_device_id: { type: DataTypes.STRING(64), unique: true },
    android_id: { type: DataTypes.STRING(64) },
    device_model: { type: DataTypes.STRING(100) },
    manufacturer: { type: DataTypes.STRING(100) },
    brand: { type: DataTypes.STRING(100) },
    device_name: { type: DataTypes.STRING(100) },
    product: { type: DataTypes.STRING(100) },
    board: { type: DataTypes.STRING(100) },
    hardware: { type: DataTypes.STRING(100) },
    build_fingerprint: { type: DataTypes.STRING(255) },
    android_version: { type: DataTypes.STRING(20) },
    sdk_version: { type: DataTypes.STRING(10) },
    fcm_token: { type: DataTypes.TEXT },
    status: enumField({ field: 'status', defaultValue: 'pending' }),
    is_online: { type: DataTypes.BOOLEAN, defaultValue: false },
    last_seen: { type: DataTypes.DATE },
    factory_reset_count: { type: DataTypes.INTEGER, defaultValue: 0 },
    last_factory_reset: { type: DataTypes.DATE },
    is_rooted: { type: DataTypes.BOOLEAN, defaultValue: false },
    safety_net_passed: { type: DataTypes.BOOLEAN },
    is_device_owner: { type: DataTypes.BOOLEAN, defaultValue: false },
    is_admin_active: { type: DataTypes.BOOLEAN, defaultValue: false },
    app_version: { type: DataTypes.STRING(20) },
    enrollment_code: { type: DataTypes.STRING(50), unique: true },
    enrolled_at: { type: DataTypes.DATE },
    enrolled_by: { type: DataTypes.STRING(36) },
    last_latitude: { type: DataTypes.STRING(50) },
    last_longitude: { type: DataTypes.STRING(50) },
    last_location_time: { type: DataTypes.DATE },
    last_location_address: { type: DataTypes.TEXT },
    camera_active: { type: DataTypes.BOOLEAN, defaultValue: false },
    last_photo_url: { type: DataTypes.TEXT },
    last_photo_time: { type: DataTypes.DATE },
    is_app_hidden: { type: DataTypes.BOOLEAN, defaultValue: false },
    is_app_disabled: { type: DataTypes.BOOLEAN, defaultValue: false },
    battery_level: { type: DataTypes.INTEGER },
    is_charging: { type: DataTypes.BOOLEAN },
    network_type: { type: DataTypes.STRING(20) },
    customer_id: { type: DataTypes.STRING(36) },
    created_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    updated_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
  },
  { tableName: 'devices', timestamps: false },
);

export { DeviceStatus };
export default Device;
